### PR TITLE
Add support for Volume FX and BASSLoud

### PR DESCRIPTION
**Pull request by smallketchup82**
Prerequisite for https://github.com/ppy/osu/pull/27793

Adds support for Volume Fx and Bassloud. Merges https://github.com/ManagedBass/ManagedBass/pull/124 and https://github.com/ManagedBass/ManagedBass/pull/121 with some custom bug fixes and patches to make them work. Also gets this branch up to date with ManagedBass master. Tested and works on MacOS and Windows

Another prerequisite for my PR is updated `ppy.ManagedBass.*` nuget packages, so if you can, please upload updated versions of the following packages if & when this gets merged:
- ppy.ManagedBass
- ppy.ManagedBass.Fx
- ppy.ManagedBass.Loud (this will also be new)

### CODE REVIEW

File: src/AddOns/BassFx/Effects/Objects/VolumeEffect.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+        /// <summary>
+        /// The curve to use in the transition... False for linear, true for logarithmic. The default value is false.
+        /// </summary>
+        public bool Curve
+        {
+            get => Convert.ToBoolean(Parameters.lCurve);
+            set
+            {
+                Parameters.lCurve = (uint)(value == true ? 1 : 0);
```
Comment (bdach) [inline src/AddOns/BassFx/Effects/Objects/VolumeEffect.cs]: redundant comparison of boolean to boolean constant

```suggestion
                Parameters.lCurve = (uint)(value ? 1 : 0);
```

File: src/Bass/Shared/ChannelReferences.cs [hunk 2]
```diff
@@ -22,7 +22,7 @@ public static void Add(int Handle, int SpecificHandle, object proc)
         {
 #if !__IOS__
             // in .NET iOS, the __IOS__ constant cannot be seen, so rely on RuntimeFeature instead.
-            if (!RuntimeFeature.IsDynamicCodeCompiled)
+            if (!RuntimeFeature.IsDynamicCodeSupported)
```
Comment (bdach) [inline src/Bass/Shared/ChannelReferences.cs]: why is this changed?

File: src/Bass/Shared/Bass/Enumerations/BassFlags.cs [hunk 3]
```diff
@@ -716,6 +716,38 @@ public enum BassFlags : uint
...
+
+        /// <summary>
+        /// BassLoud add-on: True peak level in linear scale.
+        /// </summary>
+		BassLoudnessTruePeak = 8,
+
+        /// <summary>
+        /// BassLoud add-on: Automatically free the measurement when the channel is freed.
+        /// </summary>
+		BassLoudnessAutofree = 0x8000,
```
Comment (bdach) [inline src/Bass/Shared/Bass/Enumerations/BassFlags.cs]: - broken indentation
- bit flag enums should use compile-time constant expressions with bit shifts or hex/binary literals

File: src/AddOns/Bassloud/Bassloud.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        /// <returns>If successful, the loudness measurement's channel handle is returned, else 0 is returned. Use <see cref="Bass.LastError" /> to get the error code.</returns>
+        /// <exception cref="Errors.Handle"><paramref name="Handle"/> is not valid.</exception>
+        [DllImport(DllName, EntryPoint = "BASS_Loudness_GetChannel")]
+        public static extern int BASS_Loudness_GetChannel(int Handle);
+
+        /// <summary>
+        /// Retrieves the level of a loudness measurement.
+        /// </summary>
+        /// <param name="Handle">The loudness measurement handle.</param>
+        /// <param name="Mode">The measurement type to retrieve. One of the following.</param>
```
Comment (bdach) [inline src/AddOns/Bassloud/Bassloud.cs]: incomplete xmldoc

File: src/AddOns/BassFx/Effects/Objects/VolumeEffect.cs (hunk 1 above)
Comment (smallketchup82) [reply src/AddOns/BassFx/Effects/Objects/VolumeEffect.cs]: So smoogi wants to see if we can use the [Volume channel effect instead of VolumeFx](https://github.com/ppy/osu/pull/27793#discussion_r1553041203). Depending on how that goes, this may or may not be needed. I'll wait until that gets resolved before addressing this.

File: src/Bass/Shared/ChannelReferences.cs (hunk 2 above)
Comment (smallketchup82) [reply src/Bass/Shared/ChannelReferences.cs]: It's @frenzibyte's [PR which got merged into ManagedBass](https://github.com/ManagedBass/ManagedBass/pull/118), I think it would be better to have him weigh in here. But I'd like to clarify that its in ManagedBass master.

This current PR both adds support for Volume FX and BassLoud, and also gets us up to date with ManagedBass master.

File: src/Bass/Shared/Bass/Enumerations/BassFlags.cs (hunk 3 above)
Comment (smallketchup82) [reply src/Bass/Shared/Bass/Enumerations/BassFlags.cs]: Addressed 1

The rest of the file doesn't seem to do 2, I believe that we should stay consistent. Either we change everything, or nothing.

File: src/AddOns/BassFx/Effects/Objects/VolumeEffect.cs (hunk 1 above)
Comment (smallketchup82) [reply src/AddOns/BassFx/Effects/Objects/VolumeEffect.cs]: I'll address this in a fixup PR to the PR that implemented this. I will merge that branch into this one when finished.

File: src/Bass/Shared/ChannelReferences.cs (hunk 2 above)
Comment (frenzibyte) [reply src/Bass/Shared/ChannelReferences.cs]: It seems this specific change was not included in our fork. It was discussed in https://github.com/ManagedBass/ManagedBass/pull/118#discussion_r1254046836 if you want to have a read there. 

TL;DR this change only affects osu! on release configuration with phone deployment and osu!/osu-framework on debug configuration with both phone deployment and via simulator.

I would recommend discarding this change from this PR, it's not relevant and I don't think anyone is willing to test and confirm whether it's better in practice.

Comment (smallketchup82) [reply src/Bass/Shared/ChannelReferences.cs]: Got it, I'll undo this change. We can probably merge it in at a later date.

File: src/AddOns/Bassloud/Bassloud.cs
Comment (mcendu) [inline src/AddOns/Bassloud/Bassloud.cs:1]: The naming convention used in this PInvoke interface isn't in line with the rest of ManagedBass. `BASS_Loudness_GetVersion`, for example, would probably become `BassLoud.GetVersion` or `BassLoudness.GetVersion` in the rest of ManagedBass.

Comment (smallketchup82) [reply src/AddOns/Bassloud/Bassloud.cs:1]: @DustinBond Could you comment on this?

Comment (DustinBond) [reply src/AddOns/Bassloud/Bassloud.cs:1]: I'm in agreement with @mcendu it would be better if inline with standard practices, but something is better than nothing. I would recommend an update,

Comment (smallketchup82) [reply src/AddOns/Bassloud/Bassloud.cs:1]: I've changed the methods. Dropped `BASS_Loudness` and switched to PascalCase. Let me know if it looks good

File: src/AddOns/Bassloud/Bassloud.cs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        /// </summary>
+        /// <param name="Handle">The loudness measurement handle.</param>
+        /// <param name="Mode">The measurement type to retrieve. One of the following. BassFlags.BassLoudnessCurrent, BassFlags.BassLoudnessIntegrated, BassFlags.BassLoudnessRange, BassFlags.BassLoudnessPeak, BassFlags.BassLoudnessTruePeak.</param>
+        /// <param name="Level">Pointer to a variable to receive the measurement level.</param>
+        /// <returns>If successful, TRUE is returned, else FALSE is returned. Use <see cref="Bass.LastError"/> to get the error code.</returns>
+        /// <exception cref="Errors.Handle"><paramref name="Handle"/> is not valid.</exception>
+        /// <exception cref="Errors.Parameter"><paramref name="Mode"/> is not valid. If requesting a duration with BASS_LOUDNESS_CURRENT then it exceeds what has been enabled.</exception>
+        /// <exception cref="Errors.NotAvailable">The requested measurement has not been enabled.</exception>
+        [DllImport(DllName, EntryPoint = "BASS_Loudness_GetLevel")]
+        public static extern bool GetLevel(int Handle, BassFlags Mode, ref float Level);
```
Comment (hwsmm) [inline src/AddOns/Bassloud/Bassloud.cs]:
```suggestion
        public static extern bool GetLevel(int Handle, BassFlags Mode, out float Level);
```

`out` may be better since current value of `Level` doesn't get used.

Comment (smallketchup82) [reply src/AddOns/Bassloud/Bassloud.cs]: I tried `out` before I tried `ref` when I was originally writing this. Using `out` caused a bunch of errors for some reason. My guess was the way pointers and stuff are done in C. I guess I'll give it another try later on though.

File: src/AddOns/BassFx/BassFx.csproj [hunk 6]
```diff
@@ -3,7 +3,7 @@
   <PropertyGroup>
     <TargetFramework>netstandard2.1</TargetFramework>
     <AssemblyName>ppy.ManagedBass.Fx</AssemblyName>
-    <Version>3.0.0</Version>
+    <Version>2023.0.0</Version>
```
Comment (smoogipoo) [inline src/AddOns/BassFx/BassFx.csproj]: Was this intentionally changed?

File: src/Bass/Desktop/Desktop.csproj [hunk 7]
```diff
@@ -6,7 +6,7 @@
     <GenerateDocumentationFile>true</GenerateDocumentationFile>
     <AllowUnsafeBlocks>True</AllowUnsafeBlocks>
     <DefineConstants>__DESKTOP__</DefineConstants>
-    <Version>3.0.0</Version>
+    <Version>2023.0.1</Version>
```
Comment (smoogipoo) [inline src/Bass/Desktop/Desktop.csproj]: Here too.

File: src/AddOns/Bassloud/Bassloud.cs (hunk 5 above)
Comment (smoogipoo) [reply src/AddOns/Bassloud/Bassloud.cs]: I think this is fine for now if you can't get it to work, though I think `ref` and `out` should behave similarly in this case.

File: src/AddOns/BassFx/BassFx.csproj (hunk 6 above)
Comment (smallketchup82) [reply src/AddOns/BassFx/BassFx.csproj]: Internal testing, I think I forgot to revert that. I'll make sure to do that

File: src/AddOns/Bassloud/Bassloud.cs (hunk 5 above)
Comment (hwsmm) [reply src/AddOns/Bassloud/Bassloud.cs]: `out` worked fine in my environment, but anyway I'll just revert it in my PR since it's just a minimal change.

**Comment by smallketchup82** [comment]
BassLoud support has been merged into upstream https://github.com/ManagedBass/ManagedBass/pull/124

There are some differences (mainly with the function names) between the BassLoud bindings in this PR and upstream. Should we switch to match upstream or continue with our implementation?

**Comment by smoogipoo** [comment]
We should probably just use upstream.

**Comment by smallketchup82** [comment]
I'll close this PR and create a new one getting this fork up to date with upstream. There's a couple of changes I want to make to upstream as well, so I'll submit those.
